Merge HeliLand into Land

reaperrr opened the pull request:
Another part of #15803 that could be split off relatively easy because there aren't any real dependencies (although I made it depend on #16241 to keep the rebase diff small).

reaperrr committed:
Merge HeliLand into Land activity

matjaeck commented:
Needs a rebase.

tovl commented on OpenRA.Mods.Common/Activities/Air/Land.cs:
```diff
@@ -21,18 +21,33 @@ public class Land : Activity
 		{
 			target = t;
 			aircraft = self.Trait<Aircraft>();
 			this.requireSpace = requireSpace;
```
There are only three calls to `Land` left where `requireSpace = false` and those cases could (maybe should) be handled by using `ignoreActor` instead. So maybe it is a good idea to remove this parameter?

reaperrr replied:
Done.

tovl commented on OpenRA.Mods.Common/Activities/Air/Land.cs:
```diff
@@ -21,18 +21,33 @@ public class Land : Activity
+		}
+
+		public Land(Actor self, bool requireSpace, Actor ignoreActor = null)
+			: this(self, requireSpace, WDist.Zero, ignoreActor) { }
```
Four overloads seems a bit unwieldy. Maybe it is possible to use named optional arguments for `target` and `landAltitude` instead of overloads?

reaperrr replied:
`Target` and `WDist` don't really support that, but I refactored `landAltitude` to not need a parameter and removed `requireSpace` as you suggested above, which was enough to get down to two overloads.

tovl commented on OpenRA.Mods.Common/Traits/Air/Aircraft.cs:
```diff
@@ -301,14 +301,14 @@ protected virtual void Tick(Actor self)
 			// Add land activity if LandOnCondidion resolves to true and the actor can land at the current location.
 			if (!ForceLanding && landNow.HasValue && landNow.Value && airborne && CanLand(self.Location)
-				&& !(self.CurrentActivity is HeliLand || self.CurrentActivity is Turn))
+				&& !((self.CurrentActivity is Land && Info.VTOL) || self.CurrentActivity is Turn))
```
Does this have to be exclusive to VTOL craft?

reaperrr replied:
It doesn't have to, but might not work well with non-VTOLs. I removed the limitation, but fixing any bugs with non-VTOLs would likely require refactoring `Land` further to handle the case of "already at target location, but up in the air" for non-VTOLs, which I feel is a bit out of scope for this PR.

reaperrr committed:
Streamline Land activity

Removed some redundant parameters, some redundant overloads
and made Land always consider LandAltitude relative to target.

reaperrr committed:
Allow non-VTOL to LandOnCondition

Note: This might still break in unexpected ways,
since non-VTOLs implicitly rely on ReturnToBase
to calculate the approach vector for them.

pchote commented on OpenRA.Mods.Common/Activities/Air/Land.cs:
```diff
@@ -19,18 +19,27 @@ public class Land : Activity
 		bool soundPlayed;
 
-		public Land(Actor self, Target t, bool requireSpace, Actor ignoreActor = null)
+		public Land(Actor self, Target t, Actor ignoreActor = null)
```
Can we please use this opportunity to get rid of the hacky `ignoreActor` parameter too?

The simplest way is to replace `ignoreActor` with a `WVec offset` and then change `ReturnToBase` to call this as QueueChild(self, new Land(self, Target.FromActor(dest), offset), true);`.

IMO a better way is to pass the `Exit` itself, and treat it as the requested docking point -- this takes us another step towards docking logic and airfields that support multiple actors.

reaperrr replied:
> Maybe best to do the simple approach here, and then migrate to `Exit` in a future PR?

As I've mentioned before, I'm working on (or rather have pretty much finished and just wait for dependencies to get merged) a refactor of `Reservable` into a `Dock` trait that should be more suitable than `Exit`, so yeah, I'd like to leave the migration to a follow-up.

reaperrr commented:
Updated.

pchote commented on OpenRA.Mods.Common/Activities/Air/Land.cs:
```diff
@@ -42,22 +46,28 @@ public override Activity Tick(Actor self)
 			{
-				var landingCell = self.World.Map.CellContaining(target.CenterPosition);
-				if (!aircraft.CanLand(landingCell, ignoreActor))
+				var landingCell = !aircraft.Info.VTOL ? self.World.Map.CellContaining(target.CenterPosition + offset) : self.Location;
```
Why `self.Location` for VTOL aircraft but `target.CenterPosition + offset` for everything else?

I realize that this question is coming close to a (i'm sure very annoying) scope creep: this activity assumes/requires that the actor is already at the correct X,Y position where it really should be checking this and queuing a fly child activity if it isn't.  That's out of scope for this PR, but until then I don't see why the VTOL distinction here actually matters.

pchote commented on OpenRA.Mods.Common/Activities/Air/Land.cs:
```diff
@@ -42,22 +46,28 @@ public override Activity Tick(Actor self)
 					// Maintain holding pattern.
-					QueueChild(self, new FlyCircle(self, 25), true);
+					if (!aircraft.Info.CanHover)
+						QueueChild(self, new FlyCircle(self, 25), true);
```
Isn't the actor going to be in completely the wrong position to land after this?

reaperrr replied:
Probably. But in my opinion, properly fixing issues like this requires splitting (and improving) at least the 'calculate approach vector and queue move activities' part of `ReturnToBase` to a separate activity that can be used in places like this. I can try to do that before next release, but not as part of this PR.

pchote replied:
Definitely work for a future PR.

The approach vector code should move here into `Land` rather than a separate activity, since "landing" is itself a logical activity that should be treated as a single unit - we ask the actor to "land in cell X" or "dock with actor Y" and expect it to sort out the details and deal with any unexpected delays or blockages without us having to micromanage it.

Harvesters are a good parallel to this, we used to have `FindAndDeliverResources` (and `FindResources` before it) manage everything, moving the actor and then queuing `Harvest` to harvest whatever is in the same cell as the actor.  @tovl's recent PRs changed this to make `Harvest` much more logical and robust: it is given the target cell, and works out the details of moving to the cell if needed before harvesting.

pchote reviewed (commented):
Unfortunately this regresses TS carryalls quite badly.
* Carryalls can't pickup actors (fix noted below)
* Carryall "lands" at ground level after picking up the actor, pushing the cargo below ground level
* Carryall dances around instead of landing to drop the cargo (maybe related to my landing position question above?)

tovl commented:
> Carryall "lands" at ground level after picking up the actor, pushing the cargo below ground level

This is already true on bleed and will be fixed by #16417.

reaperrr commented:
> Carryalls can't pickup actors (fix noted below)

Fixed (locally).

> Carryall "lands" at ground level after picking up the actor, pushing the cargo below ground level

As @tovl said.

> Carryall dances around instead of landing to drop the cargo (maybe related to my landing position question above?)

This one I have no clue how to fix. All I can tell is that the Carryall never reaches `Wait` or `Release` state, so it seems somehow `if (self.World.Map.DistanceAboveTerrain(carryablePosition) != WDist.Zero)` always resolves to `true` for some reason, or that the Land activity itself makes the carryall take off again before we can trigger `Release`.
Could also be that inaccurate landing offset + AltitudeVelocity make it always overshoot `WDist.Zero`, but changing the check to `> WDist.Zero` didn't fix it, so that's unlikely.
In any case, I'll need help debugging this if we want this PR merged soon, I have trouble getting my head around it right now.

pchote commented:
Wrapping up these aircraft PRs is one of my priorities for the milestone, so i'll see what I can do over the next few days.

tovl commented on OpenRA.Mods.Common/Activities/Air/Land.cs:
```diff
@@ -67,23 +77,36 @@ public override Activity Tick(Actor self)
 
-			if (!soundPlayed && aircraft.Info.LandingSounds.Length > 0 && !self.IsAtGroundLevel())
+			var altitude = self.World.Map.DistanceAboveTerrain(self.CenterPosition);
+			var landAltitude = self.World.Map.DistanceAboveTerrain(target.CenterPosition + offset) + aircraft.Info.LandAltitude;
```
> Carryall dances around instead of landing to drop the cargo (maybe related to my landing position question above?)

Rebasing this on top of #16417 and replacing `aircraft.Info.LandAltitude` with `aircraft.LandAltitude` in the above line fixes this issue for me. (introducing `aircraft.LandAltitude` is also what fixed the issue of loaded carryalls landing below the terrain.)

pchote commented:
> i'll see what I can do over the next few days.

It turns out that there are actually two different issues contributing to this.  The land position isn't quite right, and the carryall seems to be blocking itself from landing. 

I believe that f650d88ddf381caad5c8f7fb962367b37110d14f should fix these, and also the pickup issue to help with testing.  You can set a large `Carryall` `LocalOffset` to show that the drop behaviour still works as expected even if the actor to drop ends up in a different cell to the carryall's center position.

Now also needs a rebase.

pchote commented on OpenRA.Mods.Common/Activities/Air/Land.cs:
```diff
@@ -19,20 +19,24 @@ public class Land : Activity
+			: this(self, t, WVec.Zero) { }
+
+		public Land(Actor self)
+			: this(self, Target.FromPos(self.Trait<Aircraft>().GroundPosition), WVec.Zero) { }
```
`GroundPosition` is only used from here, and doesn't actually depend on `Aircraft`. You can define a `static WPos GroundPosition(Actor self)` method here and then do `Target.FromPos(GroundPosition(self))` instead.

reaperrr replied:
Went with the latter.

pchote commented:
#16417 makes some big changes to the carryall, so it might make sense to get that merged before rebasing this one.

reaperrr commented:
Rebased, applied @tovl's fixes from #16509 (squashed into 2nd commit), applied `GroundPosition` change request (also squashed into 2nd commit).

tovl commented on OpenRA.Mods.Common/Activities/Air/Land.cs:
```diff
@@ -67,23 +77,36 @@ public override Activity Tick(Actor self)
 			if (d.HorizontalLengthSquared < move.HorizontalLengthSquared)
 			{
-				aircraft.SetPosition(self, target.CenterPosition);
+				var landingAltVec = new WVec(0, 0, aircraft.Info.LandAltitude.Length);
```
I think we should use `aircraft.LandAltitude` here as well.

tovl commented on OpenRA.Mods.Common/Activities/Air/Land.cs:
```diff
@@ -67,23 +77,36 @@ public override Activity Tick(Actor self)
 			}
 
-			var landingAlt = self.World.Map.DistanceAboveTerrain(target.CenterPosition);
+			var landingAlt = self.World.Map.DistanceAboveTerrain(target.CenterPosition + offset) + aircraft.Info.LandAltitude;
```
Same here.

reaperrr commented:
Updated.

pchote commented:
> Activities/Air/Land.cs(104,40): error CS1503: Argument 3: cannot convert from 'OpenRA.WDist' to 'int' [/home/travis/build/OpenRA/OpenRA/OpenRA.Mods.Common/OpenRA.Mods.Common.csproj]

reaperrr commented:
Fixed.

pchote commented on OpenRA.Mods.Common/Activities/Air/Land.cs:
```diff
@@ -42,22 +46,28 @@ public override Activity Tick(Actor self)
+			// TargetType.Invalid indicates target was an actor and either killed, removed or teleported
+			if (target.Type == TargetType.Invalid)
+			{
+				Cancel(self);
```
Calling `Cancel` and then immediately returning `NextActivity` looks suspicious. Shouldn't we be falling through to the `IsCanceling` check below so that we can remove the ground reservation before moving on?

reaperrr replied:
To be honest, it looks suspicious to me, too. Removed the Cancel call and merged this and the below 'if'.

pchote commented on OpenRA.Mods.Common/Activities/Air/ReturnToBase.cs:
```diff
@@ -225,16 +225,10 @@ public override Activity Tick(Actor self)
-				else
-					QueueChild(self, new Land(self, Target.FromPos(dest.CenterPosition + offset), true, dest), true);
+				if (aircraft.Info.VTOL && aircraft.Info.TurnToDock)
+					QueueChild(self, new Turn(self, aircraft.Info.InitialFacing), true);
```
These `Turn`s are logically part of the landing behaviour, so really belong inside the `Land` activity.

This is already done in #16509, so we don't need to address this here.

pchote commented:
LGTM otherwise, so I expect to :+1: after my question above is resolved.

pchote commented on OpenRA.Mods.Common/Activities/Air/Land.cs:
```diff
@@ -42,22 +46,21 @@ public override Activity Tick(Actor self)
+			if (IsCanceling || target.Type == TargetType.Invalid)
 			{
 				aircraft.RemoveInfluence();
 				return NextActivity;
```
This returns without putting the actor back into a consistent state - stopping the helicopter or losing the target while the land is half done will leave the helicopter sitting at an intermediate altitude.

This is already the case on bleed and I think may already be solved in #16509, so we don't need to address this here.

tovl commented:
LGTM :+1:

pchote commented:
https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/2567491dc97df37183bf89600d37374c3a1f8cf5
